network/network_test.go

```diff
@@ -632,3 +632,41 @@ func TestDialDeletesNonValidators(t *testing.T) {
 	}
 	wg.Wait()
 }
+
+// Test that cancelling the context passed into dial
+// causes dial to return immediately.
+func TestDialContext(t *testing.T) {
+	_, networks, wg := newFullyConnectedTestNetwork(t, []router.InboundHandler{nil})
+	defer wg.Done()
+	network := networks[0]
+
+	dialer := newTestDialer()
+	network.dialer = dialer
+	dynamicIP, listener := dialer.NewListener()
+	trackedIP := &trackedIP{
+		ip: dynamicIP.IPPort(),
+	}
+
+	// Asset that when the context is cancelled, dial returns immediately.
+	// That is, [listener] doesn't accept a connection.
+	network.manuallyTrackedIDs.Add(ids.EmptyNodeID)
+	ctx, cancel := context.WithCancel(context.Background())
+	cancel()
+	network.dial(ctx, ids.EmptyNodeID, trackedIP)
+
+	gotConn := make(chan struct{})
+	go func() {
+		_, _ = listener.Accept()
+		close(gotConn)
+	}()
+	select {
+	case <-gotConn:
+		require.FailNow(t, "unexpectedly connected to peer")
+	case <-time.After(1 * time.Second):
+	}
```

Comment:
>  it can result in inconsistent behavior.

Well, if the code is correct then the behavior is consistent. That is, we'll never error due to unexpected connection. It's true that it's flaky if the code is wrong and a connection is sometimes made. I don't really know how else to test this otherwise, though.

Reply:
@joshua-kim updated test